pkg/apis/core/validation/validation.go

```diff
@@ -6329,3 +6337,20 @@ func isHeadlessService(service *core.Service) bool {
 		len(service.Spec.ClusterIPs) == 1 &&
 		service.Spec.ClusterIPs[0] == core.ClusterIPNone
 }
+
+// validateUpgradeDowngradeLoadBalancerClassName validation for loadBalancerClassName in cases of user upgrading or downgrading service
+func validateUpgradeDowngradeLoadBalancerClassName(oldService, service *core.Service) field.ErrorList {
+	allErrs := make(field.ErrorList, 0)
+	// bail out early for non LoadBalancer type of service
+	if oldService.Spec.Type != core.ServiceTypeLoadBalancer || service.Spec.Type != core.ServiceTypeLoadBalancer {
+		return allErrs
+	}
+	// test if LoadBalancerClassName set before and has changed
+	if oldService.Spec.LoadBalancerClassName != nil {
+		if service.Spec.LoadBalancerClassName == nil || *oldService.Spec.LoadBalancerClassName != *service.Spec.LoadBalancerClassName {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "loadBalancerClassName"), service.Spec.LoadBalancerClassName, "may not change once set"))
```

Comment:
I'm on the fence about whether this should really be immutable. If the class name changes, is it too much to ask controller implementations to clean up those load balancers? Maybe it is :thinking:

Reply:
I will defer to @thockin on this one.

Reply:
We need to pin down the semantics - there are too many edge cases here.

Can I change it from `nil` to a value?  If so, that is identical to allowing mutations.  So I think either:

1) ONLY allow this value to be set when type becomes LB (create or update) and never thereafter

2) Allow this value to be changed, set, or unset at any time AS LONG AS type = LB

In both cases, clear it when type becomes anything except LB

Reply:
the auto-clear can be found in pkg/registry/core/service/strategy.go `dropTypeDependentFields`